Multiple Event Listener Tweaks

juanpedromoreno opened the pull request:
This PR brings the ability to override the different handlers in the Android Listeners with several overrides (for instance: `OnSeekBarChangeListener`). 

If this contribution looks good to you, in a potential next PR, we would propose replace the current `On` implementation for the equivalent `UnitOn` proposed here, and `FuncOn` for `UnitFuncOn`, even both could be named `On` and `FuncOn`. That means the `UnitOn` is equivalent to the current `On`, and in the other hand, `UnitFuncOn` is equivalent to `FuncOn`.

From 47Deg, we wish this contribution will be useful.

Please, @stanch , Could you review? Thanks!

Commit:
Support for multi event listeners

juanpedromoreno committed:
Merge branch 'master' into juanpedro-adding-multiple-handlers-event-listener

juanpedromoreno committed:
Fixes issues related to the scala 2.10 version

juanpedromoreno committed:
Optimizes the Spec imports

juanpedromoreno committed:
Adds some pieces of documentation regarding to the new Event Tweaks

stanch commented:
Hi and thanks for your contribution!

I’ve left some comments in the main commits, so here I’d just like to do a short summary.

1. I would appreciate if you could look at the [contribution section of the README](https://github.com/macroid/macroid#contributing) and adapt the commit messages accordingly — it makes it much easier to write changelogs. Thanks!
2. Could you provide more context for the `BoolOn` handler?
3. The usage of parameter groups, such as `(onProgressChanged: UnitFuncEvent)`, to indicate the desired method is very clever. In fact, it got me confused for a second :) I wonder if we can take it even further and make something super generic, like

  ```scala
  w[Button] <~ handle {
    case Event("click", v: View) ⇒
      toast("Clicked") <~ fry
    case Event("seekBarChange.onProgressChanged", view: SeekBar, progress: Int, fromUser: Boolean) ⇒
      toast(s"Progress changed to $progress") <~ fry
  }
  ```

  What do you think?

juanpedromoreno committed:
Removes BoolEvent. Renames Handlers. Hides the ***Macros object. Addresses all comments over the previous commits.

juanpedromoreno commented:
Hi @stanch, 

I didn't see your last comment over the PR :( , sorry about it. My last commit message, it doesn't fit with the desired structure.

In the last commit, I've removed the BoolEvent handler type because of it adds some confusion and it really doesn't provide anything new.

About your 3rd point, sounds good. I'm going to take a look and I'll let you know my final thoughts about it ;)  

I have to push another commit and I promise it will be with the desired format. Thanks!

juanpedromoreno committed:
= core: updates the handler type in the default handlers.
+ core: upgrades the android-sdk plugin version.
= docs: minor fixes over the documentation.

juanpedromoreno commented:
All comments has been addressed. Could you take another look when you have a chance? Thanks!

As I said in the comment above, your proposal to improve the DSL to manage the events it seems very interesting and looks so good but at this time, I think we could go ahead with this PR such it has been implemented and in the future we could improve it ;) . What do you think?

Also, `On` is equivalent to `UnitOn` and `FuncOn` is equivalent to `UnitFuncOn`. Does it make sense if we unify them in a next PR?

stanch commented:
Hey @juanpedromoreno, sorry it’s taking so long — I’ll only be able to have a thorough look at your changes over the weekend.

stanch commented on macroid-core/src/main/scala/macroid/MultiEventTweaks.scala:
```diff
@@ -0,0 +1,146 @@
+
+  type FuncHandler = String
+
+  sealed trait HandlerType
```
Can `HandlerType`, `HUnit`, `HUnitFunc` reuse `macroid.EventTweakMacros.{ ListenerType, UnitListener, FuncListener }`?

stanch commented on macroid-core/src/main/scala/macroid/MultiEventTweaks.scala:
```diff
@@ -0,0 +1,146 @@
+
+  object UnitOn extends Dynamic {
+    def applyDynamic[W <: View](event: String)(handler: Any): Tweak[W] = macro onUnitImpl[W]
+  }
```
If I understand right, this is exactly the same as the current `On` and `FuncOn`, but delegating the implementation to the new multi-listener logic, right? I would prefer to avoid having duplicating functionality in the codebase. Perhaps you could just rewire the old objects to your implementation instead?

stanch commented on macroid-core/src/main/scala/macroid/MultiEventTweaks.scala:
```diff
@@ -0,0 +1,146 @@
+
+    val Expr(Literal(Constant(e: String))) = event
+
+    onMultipleFuncImpl(c)(event)(c.Expr[UnitHandler => Any](q"(${newTermName(s"on${e.capitalize}")}: UnitHandler) => $handler"))
```
I think you can make it more readable by assigning the quasiquote to a `val` first. `UnitHandler` has to be `_root_.macroid.MultiEventTweaks.UnitHandler`, otherwise the code will not work if the user has not imported `UnitHandler` — the same goes for every use of `UnitHandler` and `FuncHandler` in the quasiqoutes below.

stanch commented on macroid-core/src/main/scala/macroid/MultiEventTweaks.scala:
```diff
@@ -0,0 +1,146 @@
+    c.Expr[Tweak[View]](q"_root_.macroid.Tweak[$tpe] { x ⇒ x.$setter(new $listener { ..$overrides })}")
+  }
+
+  def findSetter(c: MacroContext)(event: c.Expr[String], tp: c.Type) = {
```
This method is an almost verbatim copy of the code from `EventTweakMacros.onBase`. I think it would be better to reuse the same method in both cases — could be either this one or that one.

stanch commented on macroid-core/src/main/scala/macroid/MultiEventTweaks.scala:
```diff
@@ -0,0 +1,146 @@
+      }
+
+      val handlerType = typ.toString() match {
+        case s: String if s.endsWith("UnitHandler") => HUnit
```
This does not look very robust. Isn’t it possible to get and compare the actual precise type via typechecking?

stanch commented on macroid-core/src/main/scala/macroid/MultiEventTweaks.scala:
```diff
@@ -0,0 +1,146 @@
+        case s: String if s.endsWith("FuncHandler") => HUnitFunc
+      }
+
+      val on = scala.util.Try {
```
I think this also duplicates `EventTweakMacros.onBase` — would be nice to reuse the same method.

stanch commented on macroid-core/src/main/scala/macroid/MultiEventTweaks.scala:
```diff
@@ -0,0 +1,146 @@
+    (listener, newSet)
+  }
+
+  def getOverride(c: MacroContext)(on: c.universe.MethodSymbol, f: c.Expr[Any], handlerType: HandlerType) = {
```
This looks very much like `EventTweakMacros.getListener`. Could you adapt either of these two methods so that only one of them can be reused (same as the comments above)?

stanch commented on macroid-docs/guide/Tweaks.md:
```diff
@@ -115,6 +115,50 @@ A quick rundown:
+     * _Handlers_ Handlers would be all the methods to override separated by commas if more than one is listed.
+     * Each *handler* follows this structure:
+     
+     `methodToOverride: EventType => body`
```
`HandlerType`

stanch commented on macroid-docs/guide/Tweaks.md:
````diff
@@ -115,6 +115,50 @@ A quick rundown:
     ```
+  * `MultiOn` — works for the event listeners with several methods to override. 
+  
+     It must be used as follows:
````
Looks like this line and some lines below are mis-indented by one space :) Not that it changes the HTML output (I checked), but it is a bit disturbing :)

stanch commented on project/plugins.sbt:
```diff
@@ -1,3 +1,3 @@
-addSbtPlugin("com.hanhuy.sbt" % "android-sdk-plugin" % "1.3.19")
+addSbtPlugin("com.hanhuy.sbt" % "android-sdk-plugin" % "1.3.20")
```
This shouldn’t be needed after a rebase.

stanch commented:
Hi, here goes the second bunch of comments :) At the moment I’m mostly concerned about two things:

1. `UnitOn` and `FuncUnitOn` provide the same functionality as `On` and `FuncOn`. I would prefer a single implementation, otherwise it’s a bit confusing for the end-user. It could be your implementation — we can remove the old `On` and `FuncOn` (not ideal, since it would break source compatibility) or just rewire them to your macros.
2. Some parts of code are almost exactly the same as in the old implementation (e.g. `EventTweakMacros.getListener` vs `MultiEventTweakMacros.getOverride`). I think we should unify them as much as possible. Ideally the old macros would be reused where applicable; if they are not flexible enough, you can modify them. Alternatively, the old macros can be deleted and references to them can be replaced with your new ones.

By the way, just a small note — before merging I’ll kindly ask you to squash everything into a single commit with the standardized message format and rebase on top of the current master.

juanpedromoreno committed:
Merge branch 'master' into juanpedro-adding-multiple-handlers-event-listener

juanpedromoreno committed:
- core: removes UnitOn and UnitFuncOn (considered as duplicate).
+ core: rewires the current On and FuncOn to the MultiOn base implementation. Refactors all the implementations into a single one.
= docs: minor fixes over the documentation.

juanpedromoreno commented:
Hi @stanch,

I've unified all the duplicated code and functionality into a single event implementation, adding the MultiOn feature.

Please, review it again when you have a chance. Then, when it's ready to merge I will squash all these changes into a single commit with the standardised message format and rebase on top of the current master once more ;)

stanch commented:
Hi, sorry for the delay — I’ve just returned from a vacation yesterday. I’ll try to review your changes ASAP, unfortunately before leaving I upgraded IDEA and now I [can’t import the project properly](https://youtrack.jetbrains.com/issue/SCL-8690) :(
